ENH: fix str/repr for 0d-arrays and int* scalars

ahaldane opened the pull request:
0d arrays now respect the printoptions:

Before this PR:
```python
>>> np.set_printoptions(formatter={'all': lambda x: "test"})
>>> array(1)
array(1)
```

With this PR:
```python
>>> np.set_printoptions(formatter={'all': lambda x: "test"})
>>> array(1)
array(test)
```

This PR also cleans up a lot of array2string, and eliminates a lot of unnecessary evaluated code. In particular, the `IntegerFormat` and other constructors are now only called if the array has the corresponding type. Before, all the constructors for every type were called every time array2string was called.

ahaldane commented:
I've also left in a large comment with an alternate implementation for your consideration, which is not back-compatible. Currently, no user-specified format functions have access to `precision`, `suppress_small` and so on, that extra code would make them accessible.

When it comes time to merge I will remove that comment and one or two others.

eric-wieser commented:
> In particular, the IntegerFormat and other constructors are now only called if the array has the corresponding type. 

This is also fixed in #8963 - but hackily, so probably in need of a rebase

eric-wieser commented:
Actually, I think `lazy_get` is a little more opaque than the `lambda` stuff I do in #8963 - what do you think?

eric-wieser commented:
Specifically, in a2aea7757aacaa140f22910de6b81f9196a4aecc

ahaldane commented:
Oh yeah, one more thing. I deprecated the `style` argument of `array2string`.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -287,30 +303,52 @@ def _get_format_function(data, precision, suppress_small, formatter):
     elif issubclass(dtypeobj, _nt.integer):
         if issubclass(dtypeobj, _nt.timedelta64):
-            return formatdict['timedelta']
+            return _lazyget(formatter, 'timedelta', TimedeltaFormat, data)
```
I think this would be more obvious as `_lazyget(formatter, 'timedelta', lambda: TimedeltaFormat(data))`

eric-wieser replied:
In fact, could this be written as `formatter.get('timedelta') or TimedeltaFormat(data)`?

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -44,6 +66,28 @@ def product(x, y):
+             'str_kind': ['numpystr', 'str'] }
+
+    formatdict = {}
+    for tp, fmtfunc in six.iteritems(formatter):
```
Do we have `six` available? Also, `.items()` would work just fine here

ahaldane commented:
Yeah the lambda version is pretty nice. 

Maybe we can combine the lambdas for the raw types, plus the `_expand_formatter_kinds` from here. 

(Though now that I think about it, my `_expand_formatter_kinds` slighly changes the behavior since now the `*kinds` don't show up in `get_printoptions`)

eric-wieser commented:
Do you want to split your PRs one last time, and make the `arrayprint` fixes without the 0d case?

ahaldane commented:
Sure, let me review your PR.

ahaldane committed:
ENH: str/repr fixed for 0d-arrays

0d arrays now use the arrayprint.py formatters to print themselves.
Deprecates 'style' argument to array2string. Integer scalars are no
longer printed using the function set with ``np.set_print_function``.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -373,29 +374,24 @@ gentype_nonzero_number(PyObject *m1)
+        return NULL;
     }
-    return ret;
+    return PyObject_CallFunction(fallback, "O", self);
```
What happens if we leave this as it was before?

ahaldane replied:
I haven't actually checked, but I am 95% sure we will get infinite recursions because of tests which do things like

```python
>>> np.set_printoptions(formatter={'all':lambda x: str(x)})
```

eric-wieser replied:
> we will get infinite recursions

Hopefully they should now be bounded - that was why wanted that other commit merged

That doesn't seem sane as a formatter though anyway - "Changing the `__str__` for arrays to call `str(self)`" is obviously very wrong

ahaldane replied:
#8963 it doesn't protect against this kind of recursion (I noted that there).

If we disallow this kind formatter, though, it will be a backwards-compatibility break, which I was trying to avoid.

This formatter makes sense if you assume a clear division between array reprs and scalar reprs. If you assume that scalar reprs are totally distinct from arrays, and x is a scalar, then there can't be any recursion problem. 

The problem only arises because we happened to implement a small number of scalar reprs by calling promoting to 0d and calling the array repr, which can be interpreted as an implementation detail.

ahaldane replied:
Also, note the comment I added at the top of the file:

Both `scalartypes.c.src` and `arrayprint.py` implement reprs for scalars, but for different purposes. `scalartypes.c.src` has reprs for when the scalar is printed on its own, while `arrayprint.py` has reprs for when scalars are printed inside an ndarray.

This is quite apparent in datetime64, for example:
```python
>>> a = np.datetime64('2005-02-25')
>>> a
numpy.datetime64('2005-02-25')
>>> np.array([a])
array(['2005-02-25'], dtype='datetime64[D]')

eric-wieser replied:
>  protect against this kind of recursion (I noted that there).

You noted that it doesn't protect again `str(np.array(x))` - although you're right, I guess that is what is happening here.

> This is quite apparent in datetime64, for example:

eric-wieser replied:
(And as a follow up, can a test be added for that new repr)

ahaldane replied:
It gives
```python
>>> repr(np.datetime64('2005-02-25')[...])
"array('2005-02-25', dtype='datetime64[D]')"
```

(it used to give `"array(datetime.date(2005, 2, 25), dtype='datetime64[D]')"`)

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -496,22 +520,32 @@ def array2string(a, max_line_width=None, precision=None,
+        format_function = _get_format_function(
+                a.reshape((1,)), precision, suppress_small, formatter)
+        lst = format_function(a[()])
+    elif reduce(product, a.shape) == 0:
```
Needs a rebase, I think - this won't work well with changes merged from #8963

ahaldane commented:
Updated with only the 0d array fix.

I split off my arrayprint cleanup to a branch in my repository (here)[https://github.com/ahaldane/numpy/commit/d9d5c6fc32996a552a348e9b61910f87457d1e55]. I can still submit it if anyone wants, but it's not really necessary because of #8963.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -496,22 +520,32 @@ def array2string(a, max_line_width=None, precision=None,
-    elif functools.reduce(product, a.shape) == 0:
+        format_function = _get_format_function(
+                a.reshape((1,)), precision, suppress_small, formatter)
+        lst = format_function(a[()])
```
This is starting to feel like the job of `_array2string`, in an `if rank == 0` case

ahaldane replied:
I think you mean `_formatArray`, but yes

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -373,29 +374,24 @@ gentype_nonzero_number(PyObject *m1)
+        return NULL;
     }
-    return ret;
+    return PyObject_CallFunction(fallback, "O", self);
```
Proposal: return `repr(item())` here instead? Since this is a fallback anyway...

Not that I disagree with how you currently have this.

ahaldane replied:
what about something like

`"{name}({})".format(x.dtype.name, str(x))`

eric-wieser replied:
Ooh, that's pretty nice - with `x = self.item()`, I assume.

Also, I think that should always call `__repr__`

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -373,29 +374,24 @@ gentype_nonzero_number(PyObject *m1)
-
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        /* XXX: Why are we using str here? */
```
This comment needs to be kept in some form - because we still have the same code path for both `str` and `repr`

ahaldane commented:
Updated as discussed. In the end I made `gentype_repr` print

```python
x.dtype.type.__name__ + "(" + repr(x.item()) + ")"
```

(`x.dtype.name` was no good because it gives values like `void48` for flexible types)

Currently this gives something a little weird for voids types: `repr(void(4))` gives `void(array([0, 0, 0, 0, 0, 0], dtype=int8))`. But this should be fixed soon by some combination of #8157, #8981.

ahaldane commented:
Oh right... I need to implement reprs for all the integer types to avoid printing `int64(4)` everywhere. A of the integer types use `gentype_repr`. I'll get back to this later.

eric-wieser commented:
> . I need to implement reprs for all the integer types to avoid printing int64(4) everywhere

Actually, this seems pretty desirable to me? Not for `str` as that would break formatting - but for `repr`, that actually gives the user helpful information

ahaldane commented:
It's not a bad idea, but that may be too big a change: The python and ipython shells print the repr of the last line. So every time you do something like `>>> np.sum(arr)` it would print `int64(100)` instead of just `100` as it does now. In other words, about half of all common numpy commands would print differently.

ahaldane commented on numpy/core/arrayprint.py:
```diff
@@ -798,22 +832,21 @@ def __call__(self, x):
-                # data contains a NaT
-                max_str_len = max(max_str_len, 5)
-            self.format = '%' + str(max_str_len) + 'd'
-            self._nat = "'NaT'".rjust(max_str_len)
```
I just removed the if-statement and unindented here. No code changes

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -373,31 +374,62 @@ gentype_nonzero_number(PyObject *m1)
+    if (dtype_name == NULL) {
+        /* this should never happen since types always have __name__ */
+        PyErr_Format(PyExc_RuntimeError, "missing __name__ attribute");
+        return NULL;
```
CPython just leaves the error as is internally: https://github.com/python/cpython/blob/ba85d69a3e3610bdd05f0dd372cf4ebca178c7fb/Python/_warnings.c#L315

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -373,31 +374,62 @@ gentype_nonzero_number(PyObject *m1)
+
+    PyUString_ConcatAndDel(&dtype_name, PyUString_FromString("("));
+    PyUString_ConcatAndDel(&dtype_name, item_repr);
+    PyUString_ConcatAndDel(&dtype_name, PyUString_FromString(")"));
```
I think this might need error checking?

ahaldane replied:
`PyUString_ConcatAndDel` does all error checking for us, according to docs and to the fact that it is used this way in other places in numpy.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -373,31 +374,62 @@ gentype_nonzero_number(PyObject *m1)
+    item_repr = PyObject_Str(item);
+    Py_DECREF(item);
+    return item_repr;
+}
```
I think this should also be what `gentype_str` does? Is there any reason for ints to not pass through `_scalar_fallback`, when everything else does? (or vice versa)

ahaldane replied:
I guess our intention is slightly more explicit (doesn't go to the 'last resort'), and it's a slight optimization since we avoid calling `get_format_function`.

I think that with this only void (and user-defined) types use the scalar fallback.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -482,6 +500,11 @@ def array2string(a, max_line_width=None, precision=None,
+    # Deprecation added Apr 2017,  v1.13
+    if style is not None:
+        warnings.warn("'style' argument is deprecated and no longer functional",
+                      DeprecationWarning, stacklevel=3)
```
Why level 3 here and not 2?

ahaldane replied:
Because of you recursion wrapper :)

eric-wieser replied:
Ah, yep - that would do it! Forgot how decorators worked for a moment there

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -4226,6 +4256,17 @@ initialize_numeric_types(void)
+     */
+
+    Py@Type@ArrType_Type.tp_str = genint_type_str;
+    Py@Type@ArrType_Type.tp_repr = genint_type_str;
```
This can go inside the above repeat, right?

ahaldane replied:
I find it cleaner as a separate repeat, since it is only 2 lines. This way we avoid nested repeats and the need for `repeat1`, `repeat2`.

homu commented:
:umbrella: The latest upstream changes (presumably #9015) made this pull request unmergeable. Please resolve the merge conflicts.

ahaldane commented on numpy/core/arrayprint.py:
```diff
@@ -507,14 +530,9 @@ def array2string(a, max_line_width=None, precision=None,
-            lst = style(x)
+        format_function = _get_format_function(
+                a.reshape((1,)), precision, suppress_small, formatter)
+        lst = format_function(a[()])
```
You previously suggested this case could be moved to `_array2string` in an `if rank == 0` case, but I like it here to emphasize it is treated specially.

eric-wieser replied:
In what sense is this special though?

ahaldane replied:
OK I looked a bit more at this, and I remember another problem that makes me want to keep it this way: There is an cyclic import problem that seems complicated to fix.

Below is a traceback if I remove this special case which occurs when I try to import numpy.

The problem is that numpy `core` modules (specifically `getlimits`) want to call `array2string` before all numpy submodules are loaded. However `array2string` needs to be able to call `np.ravel` which checks for `np.matrix`, which hasn't been loaded into numpy at that point, so fails.

I don't see any simple fix to this involving changing the import order of the submodules, since `matrix` itself depends on `getlimits`, so there is a true cyclic import problem there.

This all sounds like something I'd rather work around in this PR. Perhaps the fix to to make `getlimits` not call `array2string`, or perhaps `array2string` not depend on `ravel`, I don't know, but in any case it seems fairly unrelated to 0d reprs.


```
Traceback (most recent call last):
  File "runtests.py", line 469, in <module>
    main(argv=sys.argv[1:])
  File "runtests.py", line 179, in main
    import numpy as np
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/__init__.py", line 142, in <module>
    from . import add_newdocs
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/add_newdocs.py", line 13, in <module>
    from numpy.lib import add_newdoc
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/lib/__init__.py", line 8, in <module>
    from .type_check import *
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/lib/type_check.py", line 11, in <module>
    import numpy.core.numeric as _nx
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/__init__.py", line 51, in <module>
    from . import getlimits
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/getlimits.py", line 163, in <module>
    tiny=_tiny_f128)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/getlimits.py", line 82, in __init__
    self._str_eps = float_to_str(self.eps)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/getlimits.py", line 71, in <lambda>
    float_to_str = lambda v: (params['fmt'] % array(_fr0(v)[0], ftype))
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/numeric.py", line 1939, in array_str
    return array2string(a, max_line_width, precision, suppress_small, ' ', "")
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/arrayprint.py", line 409, in wrapper
    return f(self, *args, **kwargs)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/arrayprint.py", line 537, in array2string
    separator, prefix, formatter=formatter)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/arrayprint.py", line 372, in _array2string
    data = ravel(asarray(a))
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/fromnumeric.py", line 1481, in ravel
    if isinstance(a, np.matrix):
AttributeError: 'module' object has no attribute 'matrix'
```

ahaldane commented:
Rebased and tweaked the note. I think I addressed all the recent comments.

By the way, I think the only numpy scalar type that currently uses the `gentype_repr` is `np.void`, and in that case gives a slightly wonky repr (eg `void(array([0], dtype=int8))` in python2), but that will be fixed soon in #8981.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -523,6 +541,21 @@ def array2string(a, max_line_width=None, precision=None,
+
+    # as a last resort (no implementation found) use the scalar's .item()
+    if format_function is repr_format:
+        return str(a.item())
```
Sure we don't want `repr` here?

ahaldane replied:
This fallback is for `gentype_str`, not for `gentype_repr`, so doesn't string make more sense?

eric-wieser replied:
In which case, this should probably be renamed to `_scalar_str_fallback`

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -15,6 +15,26 @@
+# printed inside an ndarray.
+#
+# Not all scalars have repr implementations in scalartypes.c.src, and if they
+# don't they will call _scalar_fallback in this file to print the scalar.
```
But here you say that `_scalar_fallback` is a fallback for `repr`

ahaldane commented:
All right, updated the `scalar_str_fallback` name and made the top comment much more precise and accurate, as you rightly suggested.

ahaldane commented:
Rebased again. I also moved the release note to 1.14.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -506,16 +521,7 @@ def array2string(a, max_line_width=None, precision=None,
-        else:
-            lst = style(x)
-    elif functools.reduce(product, a.shape) == 0:
+    if a.shape != () and functools.reduce(product, a.shape) == 0:
```
I think this entire condition can become `if a.size == 0`

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -353,6 +353,21 @@ gentype_repr(PyObject *self)
+        return NULL;
+    }
+
+    item_str = PyObject_Str(item);
```
Now that this is much simpler, can we add a `genint_type_repr` that does the same thing but calls `PyObject_Repr`, for completeness?

eric-wieser replied:
Oops, I forgot this was for all types, not just int types types. Guess it doesn't do any harm though (and on python 2, makes a distinction between longs and ints)

ahaldane replied:
Hmm maybe that's bad though. I want to preserve the old int-printing behavior as much as possible.

I don't have a 32-bit system to test on, but this means `repr(np.int64(2**63-1))` will now have an extra `L` on that arch, right ?

ahaldane replied:
OK, yeah it does. I'd like to to remove the `genint_repr` here, if you agree.

eric-wieser replied:
Yep, it will show that `L`, which you already see with `np.int64(1).item()`.

So yeah, undo that suggestion - sorry. Probably worth adding a comment to where `repr` is set then, indicating that we stick with str to avoid the trailing `L`.

Also, re commit message - we're not really "fixing" int scalars here, are we?

ahaldane commented:
Rebased now that #9113 is in.

Also, I decided to remove all the changes to `gentype_repr` and `gentype_str` and the function `scalar_str_fallback` (in the extra trial commit), as well as a large part of the comment I had written. That makes this PR much simpler and more conservative.

I did this because no numpy types will actually use those three functions after the addition of `genint_repr` in this PR and the proposed changes for void types in #8981. I can't even write tests for these functions since no code paths use them, so it seems pointless to start changing them.

(Conceivably, someone out there has written a custom dtype that uses those code paths, in which case we might as well leave it alone to avoid surprising them.)

The point of this PR is to prepare for #8981 and #8157, and the `gentype_str/repr` changes are unnecessary for that purpose.

eric-wieser reviewed (commented):
This looks way simpler than what was here before - just a few nitpicks

ahaldane commented:
Updates with your suggestions.

Once it looks good I'll squash.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -4185,6 +4214,18 @@ initialize_numeric_types(void)
+
+    /**end repeat**/
+
+
```
Too many blank lines? The rest of this block is all single blanks. Not important though

eric-wieser reviewed (approved):
Looks good to me

ahaldane commented:
updated with a comment about repr and trailing `L`, and updated the commit message

eric-wieser commented:
> scalar str/repr implemented for integer types.

Isn't a better description that integer scalars can no longer have their repr overridden?

ahaldane commented:
That's true, I hadn't thought much about that. Is that a back-compatibility problem, actually?

Maybe we do need a callback to `arrayprint.py`, but inside `genint_type_str`.

eric-wieser commented:
Could integers be overriden anyway though? I think we just fell back on calling `str/repr` on their `.item` anyway

ahaldane commented:
Oh right. The old behavior was to promote to a 0d array, which then had a special case using `.item()`.

So I think the correct summary is that printing of integer types is unchanged (for the user).

eric-wieser commented:
Sorry, with `np.set_string_function`

eric-wieser commented:
Huh, I can't seem to make that affect scalar printouts either, although I'm not sure why

ahaldane commented:
I think the "normal" chain of events when you call `repr` on an ndarray is that it calls the C-function `array_repr` in `arrayobject.c`, which calls a python function `array_repr` in `numeric.py`, which calls `array2string`. 

`np.set_string_function` allows the user to override the choice of `numeric.array_repr` in the second step, while `np.set_printoptions` controls how `array2string` works.

So in all of this, the `repr`/`str` of numpy scalars are totally unaffected.

eric-wieser commented:
Right, but didn't calling repr on an int scalar fall back to calling `ndarray.repr`, ie go through the `set_string_function` path before? Whereas now, it does not go through that path.

I can't construct a testcase though.

ahaldane commented:
I see. I can get it to work with numpy master:

```python
>>> np.set_string_function(lambda x: "test", 0)
>>> np.int32(8)
test
```

ahaldane commented:
added a note about the change, although I'd be pretty surprised it affected anyone!

eric-wieser commented:
Ah, my mistake was setting the `repr`, but I forgot that we fall back on `array_str`.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -28,12 +35,14 @@
     except ImportError:
         from dummy_thread import get_ident
 
+import numpy as np
```
This import seems iffy...  Is this sensible from a cyclic import perspective?

ahaldane replied:
I assumed it was OK since `fromnumeric.py` does it too, but I agree `from .._globals import _NoValue` seems less risky.

eric-wieser replied:
You're right, there's definitely precedent for how you had this before. Maybe leave it as it was, and we can deal with `_globals` in general at a later point (it needs the new exception types too).

ahaldane replied:
Fair enough, done.

eric-wieser commented:
Think this is all set. Thanks @ahaldane!

ahaldane commented:
Great reviewing, @eric-wieser

ahaldane commented:
Just want to note the consequences of this PR can be seen in  https://github.com/astropy/astropy/pull/6090. It looks like some of their string representations have changed slightly.

mhvk commented:
Indeed, this caused quite a few test failures in astropy; these things are somewhat of a pain with doctests. We had a similar problem when the PR that made structured arrays respect the print options was merged, but, like for that one, I think it is well worth it: scalars should not be typeset differently from arrays!

ahaldane commented:
Looking at the fixes, I actually don't very much ike the extra spaces this PR added in many arrays right after the `(` (which curiously didn't show up in any numpy tests):
```python
>>> np.array(1.0)
array( 1.)
>>> np.array(True)
array( True, dtype=bool)
```
(no spaces are added for integer types though)

I kind of want to fix that... I'll take a look.

ahaldane commented:
Does anyone know why the default is to print float arrays with two spaces? Ie, 
```python
>>> np.arange(10.)
array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9.])
``` 
instead of
```python
>>> np.arange(10.)
array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
```
?

It's easy to remove the space (just change a 2 to a 1 [here](https://github.com/numpy/numpy/blob/beac50cf98f450539dcdeee0273cfe5175d45d26/numpy/core/arrayprint.py#L660)) and the few tests that break don't seem to explain the double space.

mhvk commented:
I think it just leaves space for the sign, i.e., if you have negative numbers in your example, things will line up. For the same reason, `False` does not get an extra space, unlike `True`:
```
In [2]: np.array(True), np.array(False)
Out[2]: (array( True, dtype=bool), array(False, dtype=bool))
```
Now, arguably, since the array printing code already looks at the number of trailing digits that are required, it might as well also look at whether the space for the sign is required.

mhvk commented:
@ahaldane - do you think you are going to change it? I have an astropy PR that changes our tests for the new repr (as we try to ensure numpy-dev passes our tests), which I will hold off on if you are. For what it is worth, my sense would be to remove the extraneous space if there are no negative numbers (which means more changes in astropy, but so be it...) and similarly for `True`.

ahaldane commented:
Yeah I'd like to try, it doesn't look too hard.

I'll try to get something together later today or tonight.

eric-wieser commented:
I might be tempted to leave the `True/False` padding in all but the scalar case, as there's exactly one array for a given shape that  does not need the padding - it's a lot easier to compare things when they align:
```
>>> a
array([ True,  True,  True])
>>> b
array([False,  True,  True])
```

mhvk commented:
True, though that same argument would suggest to keep the space for the sign (which it currently does not only for float!?)
```
In [3]: np.arange(0, -10, -1)
Out[3]: array([ 0, -1, -2, -3, -4, -5, -6, -7, -8, -9])

In [4]: np.arange(0, 10, 1)
Out[4]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [5]: np.arange(0, 10., 1)
Out[5]: array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9.])

In [6]: np.arange(0, -10., -1)
Out[6]: array([ 0., -1., -2., -3., -4., -5., -6., -7., -8., -9.])

eric-wieser commented:
Arguably, yes. You could also use the same argument to pad for extra digits, to allow comparison with arrays with larger numbers.

To me, it seems to be about breaking the format into subclasses, such that within each class, the formats are consistent. Right now, these classes are:
 * one digit all positive ints
 * one digit ints
 * boolean
 * ...

Splitting `bool` into two categories is useless, because one of those categories contains exactly one member, so has no format to be consistent with.

charris commented:
Note that this has led to failing tests in scipy, see scipy/scipy#7418.

WarrenWeckesser commented:
These changes might cause issues in downstream projects.  The change in `str(array('foo'))` appears to be affecting scipy (but that hasn't been confirmed yet).

With numpy 1.12.1:
```
In [141]: np.__version__
Out[141]: '1.12.1'

In [142]: str(np.array('foo'))
Out[142]: 'foo'

In [143]: str(np.array(np.pi))
Out[143]: '3.141592653589793'
```

With numpy master:

```
In [10]: np.__version__
Out[10]: '1.14.0.dev0+1ec9ad6'

In [11]: str(np.array('foo'))
Out[11]: "'foo'"

In [12]: str(np.array(np.pi))
Out[12]: ' 3.14159265'
```
There are now quotes included in `str(array('foo'))` (as if it was using `repr()` instead of `str()`), and the format for a floating point value has fewer digits and includes a leading space.

ahaldane commented:
I do think that in principle this PR and #9139 are more correct behavior, but I recognize it is annoying for everyone to update the spaces in all their doctests. I'm open to reverting or leaving them for a more future release.

Here's a more conservative change I am thinking about. Note how:
  * The old behavior for 0d arrays was to print `str(x.item())`. 
  * This PR's behavior is to print `arrayprint.formatter(x)`.
  * A third unexplored option is to print `str(x)`, ie rely on the string functions in `scalartypes.c.src`.

In other words, in the third option we could keep the new implementations of `genint_type_str` in this PR and the proposed void-reprs in #8981. This avoid using `str(item())` which gives weird behavior for types whose `.item()` method gives something strange, but also (I think) avoid large changes to doctests of this PR.

I don't know all the implications of the third option yet.

eric-wieser commented:
I think that we need `str(x)` should be mapped to `str(x[()])` for 0d arrays. `str` isn't just used for `print`, it's used to convert numpy string-likes into python strings.

And really, we want 0d arrays to match the behaviour of scalars as closely as possible.

mhvk commented:
@eric-wieser - I'm not sure I agree - at least in my ideal future, numpy scalars don't exist at all, one just has array scalars and these behave the same way as regular arrays. (I think many problems have arisen from a poor analogy between arrays and lists.)

ahaldane commented:
I'm in the middle of trying option 3 out in #9143. (I might not have time to work on it more today).

mhvk commented:
It may be good to try define a final goal. My feeling would be that any array, scalar or otherwise, should be typeset by `arrayprint.formatter`. However, I think we do need to normalize away an extra space if all values are positive, and if there is a single `True`.

mhvk commented:
See #9144 for a quick check what happens if one removes extraneous spaces for positive numbers (and all-True booleans).
